### ISSUE TITLE
WCAG F85: Clarify guidance about where the focus should go when the modal dialog is closed

### ISSUE THREAD

**Issue by iamrafan** [w3c/wcag#518]
The [keyboard interaction for modal dialog](https://www.w3.org/TR/wai-aria-practices/#issue-container-generatedID-7) says: when a dialog is closed, the focus can be set on another element that provides logical work flow

![image](https://user-images.githubusercontent.com/22068434/46833578-cd035380-cd5d-11e8-877a-c4a7260b45b9.png)

But 

The [common failure technique F85](https://www.w3.org/WAI/WCAG21/Techniques/failures/F85) says it is a WCAG failure if the focus does not go back to the trigger control or the control next to it when the dialog is closed

![image](https://user-images.githubusercontent.com/22068434/46833445-89104e80-cd5d-11e8-82c7-cae5e4cdc1d0.png)

One example where the WCAG would fail is:

>A cancel button which opens a confirmation dialog. This cancel button would likely take to a screen where that cancel button is no longer present.

**Comment by awkawk** [comment]
I think that these at slightly different. The F85 failure is all about focus being sequential to the triggering item and part of the test involved dismissing the dialog (cancelling it). The ARIA advice seems to include guidance when executing some function in the dialog and doesn't specifically address the dismiss case. @jnurthen - do you agree?

**Comment by jnurthen** [comment]
@awkawk I see what you are saying - dismiss the dialog means "cancel" without doing any action. OK - I can buy that.
I read "dismiss" as do any action which causes the dialog to close, not just the cancel operation.

**Comment by awkawk** [comment]
Right. I think that was done because often people bother to move the focus to the dialog when it is triggered but don't restore the focus if you cancel it, and both are needed.

Perhaps we should clarify that dismiss means to cancel without performing any action in F85?  @iamrafan does that make sense to you?

**Comment by iamrafan** [comment]
@awkawk Yes, it makes sense. But, don't you think the focus should return to only the trigger control rather than either trigger or the control next to the trigger as mentioned in the test procedure in F85

> 
![image](https://user-images.githubusercontent.com/22068434/48336706-c9d9dc80-e615-11e8-8b9b-2fd44590a7a3.png)

**Comment by awkawk** [comment]
@iamrafan good point - I'm not sure what cases might explain that. @jnurthen - any thoughts?

**Comment by jnurthen** [comment]
The 2nd bullets on both (1) and (2) allow someone to simply show a pre-defined dialog which is in the HTML at the correct point and then hide it when cancelled. While it does not provide ideal focus management it does not break things horribly. I imagine this is why this is allowed.
This allows someone to create dialog-like functionality without having to worry about focus management as much.

(Note - I have never done this, but it seems like a logical reason to allow it)

**Comment by electronicwoft** [comment]
what about for dialogs that have no trigger point such as a session timeout? these typically need to be focused to act upon. Where is focus returned to once a session is resumed?

**Comment by jake-abma** [comment]
@electronicwoft the focus was somewhere already before triggering the dialog, placing it back to that point would be best. Otherwise the start of the page would be the best alternative. Or do you see other options?

**Comment by electronicwoft** [comment]
no, as a screen reader user, I find it immensely irritating if focus returns to the document once something like a dialog is actioned, but, in all these cases, there is a requirement for author-provided focus management because - in my experience at least - the code for modal/modeless dialogs is rarely adjacent to its trigger point. Also, in the case of dropdown menus, it means that tabbing becomes the de facto method for navigating a menu which is not always appropriate (i.e., mega menus composed of endlessly nested lists) or operating-system styles of menus. This means that the vast majority of web sites and applications are (and do) likely to fail 2.4.3, but, as far as I am aware, there is no sufficient technique that provides guidance on implementing this kind of focus management without trigger points etc. On another tangent, the language in the description is gender-specific such that a 'user' is referred to with 'he' and 'his' ...

**Comment by fstrr** [comment]
Since this issue was opened nearly four years ago, we've finally got the native HTML `dialog` element working in what are now the major browsers (Chromium, Safari, Firefox). And there was much rejoicing. I've put some tests together using the `dialog` element and the standard `showModal`, `show`, and `close` calls to open and close the element. These have been tested in versions of Chromium, Safari, and Firefox that support the `dialog` element.

## Test 1: opening a modal dialog by clicking on a `button`
[Testing opening a modal dialog by clicking on a `button`](https://codepen.io/fstrr/full/vYdVKVX) shows that browsers handle focus management for you, including putting focus onto the first focusable item inside the `dialog`, trapping focus, and placing the focus back onto the triggering control when the `dialog` is dismissed (by pressing the <kbd>Escape</kbd> key) or closed (by pressing the Close button).

## Test 2: the browser automatically opens a modal dialog
[Testing the browser automatically opening a modal dialog](https://codepen.io/fstrr/full/abqmqYJ) shows that browsers handle focus management for you, including putting focus onto the first focusable item inside the `dialog` and also placing the focus back onto the control that last had focus on the underlying page, even though it wasn't the triggering control. This isn't relevant to F85 as there is no trigger control, but it was mentioned in this thread, so I tested it.

## Test 3: removing the triggering element after the modal `dialog` is opened
[Testing removing the triggering element after the modal `dialog` is opened](https://codepen.io/fstrr/full/vYdVXjd) shows that all three browsers lose track of what was focused before the now-disappeared control. With the triggering component no-longer on the page, pressing the <kbd>Tab</kbd> key starts puts the user's focus into the browser's address bar, and pressing <kbd><kbd>Shift</kbd> + <kbd>Tab</kbd></kbd> puts the user's focus onto the last interactive control on the page. This is the same as [Example 2 in F85](https://www.w3.org/WAI/WCAG21/Techniques/failures/F85#example-2-setting-focus-to-the-document-after-dismissing-a-menu-embedded-on-the-page).

## Test 4: testing a non-modal `dialog`
[Testing a non-modal `dialog`](https://codepen.io/fstrr/full/vYdVxoa) shows that when a non-modal `dialog` is opened, all three browsers move the user's focus into it. Additionally, all three browsers don't allow the <kbd>Escape</kbd> key to be used to close the non-modal dialog, but when the `dialog`'s Close `button` is pressed, focus is returned to the triggering control.

<hr>
As an additional point on this thread: there seems to be a loophole in the tests, which say "check whether advancing the focus backwards in the sequential navigation order" and "check whether advancing the focus in the sequential navigation order". What's not specified is how many times is okay to advance the focus. If I have to press the <kbd>tab</kbd> key 20 times, is that okay?

## Proposed response:

### Add this sentence to the end of the Description:
Dismissing a dialog or menu means to close it without taking any action on it. For example: a modal dialog opens on a Web page and displays a sign-up form for a mailing list. A user dismisses the dialog by either pressing a "close" button or using the Escape key.

### Re-write the tests to say:
1. Activate the trigger control via the keyboard.
    - Check whether focus is in the menu or dialog.
    - Check whether advancing the focus <b>once</b> in the sequential navigation order puts focus in the menu or dialog.
2. Dismiss the menu or dialog
    - Check whether focus is on the trigger control.
    - Check whether advancing the focus <b>once</b> backwards in the sequential navigation order puts focus in the trigger control.

**Comment by alastc** [comment]
> The APG guidance allows more flexibility for the author to move focus back to a logical place rather than always to the triggering element. IMO F85 is too restrictive for a Failure technique.

Does there need to be a third 'get out clause', where the focus is placed in a logical (but not defined by us) place?
